mantidproject/mantid
# Fix PointData/Histogram conversion of RebinnedOutput

**Description of work.**



This PR makes the routines which convert to/from histogram/point data understand the `RebinnedOutput` type which carries bin fractions as well as bin frequencies. Now when these are converted to point data the fractions are set to unity.

It also reverts `SofQWMoments` to the original code/flow, and fixes the changes to `Integration` for the `RebinnedOutput` types introduced in #33408 . `SofQWMoments` relies on multiplication of the signals as points rather than bins and hence converts the data from histograms to points for a some of its processing. During this conversion, because `XDataConverter` is not aware of the bin fractions arising from the fractional rebinning algorithm `SofQW3` which produces the `RebinnedOutput` type it discards them (sets them to zero) which then causes the new `Integration` routine which _is_ aware of them to output `NaN`s because it uses the fractions to normalise the output data (as described in the issue #33407 which PR #33408 fixes).

`RebinnedOutput` types are only created by `SofQW3` and this output is always a histogram. Hence any call to `XConvertData` on this is either: 1) from `ConvertToPointData` called on the output of `SofQW3` or 2) `ConvertToHistogram` called on the output of `ConvertToPointData` from the output of `SofQW3`. Since fractional bin weights make no sense when we are discarding the bin information in `ConvertToPointData`, this PR sets all the fractional weights to unity (which effectively restores the behaviour of `Integration` to what it was before PR #33408).

The PR also reverts all the changes in PR #33798 (including changes to the test data) which was a Python-only fix of this issue by removing some of the `ConvertToPointData` calls - however it seems that these are necessary for `SofQWMoments` to function.

**To test:**

Make sure unit tests of `SofQWMoments` pass. Run `SofQWMoments` on some test data and make sure the output matches that from Mantid 6.2 or earlier versions.



Fixes #33937. 




Release notes and other documentation to follow.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
